### ISSUE TITLE
loading profile from command line argument `-p` appears to ignore ff_gain value

### ISSUE THREAD

**Issue by kevenwyld** [berarma/oversteer#93]
I have the following profile located at `$HOME/.config/oversteer/profiles/windows_defaults.ini` for my t300rs saved using the gui:

```
[DEFAULT]
range = 1080
ff_gain = 75
autocenter = 0
spring_level = 100
damper_level = 100
friction_level = 100
range_overlay = never
use_buttons = 0
center_wheel = 0
start_app_manually = 0
```

When I try to load it with `oversteer -p windows_defaults` the `range` value is set correctly but the `ff_gain` value remains where it was (at 100 in this case).

Sorry if I'm being dumb or something....  maybe I'm just doing something wrong.

**Comment by kevenwyld** [comment]
I also tried specifying the device path as suggested in a closed issue I found.  That didn't seem to make a difference.  I only have one wheel anyway.

**Comment by kevenwyld** [comment]
Also tried running it with `--debug`.  The output has some strange values for FF gain:

```
DEBUG:root:update_device_list: input-pci-0000_00_14_0-usb-0_3_2_4_1_0
DEBUG:root:update_device_list: input-pci-0000_00_14_0-usb-0_3_2_4_1_0
DEBUG:root:update_device_list: input-pci-0000_00_14_0-usb-0_3_2_4_1_0
DEBUG:root:Devices: {'input-pci-0000_00_14_0-usb-0_3_2_4_1_0': <oversteer.device.Device object at 0x7f82e19879a0>}
DEBUG:root:Setting range: 1080
DEBUG:root:Setting autocenter strength: 0
DEBUG:root:Setting FF gain: 65535
DEBUG:root:Setting spring level: 100
DEBUG:root:Setting damper level: 100
DEBUG:root:Setting friction level: 100
DEBUG:root:Setting range: 1080
DEBUG:root:Setting autocenter strength: 0
DEBUG:root:Setting FF gain: 49151
DEBUG:root:Setting spring level: 100
DEBUG:root:Setting damper level: 100
DEBUG:root:Setting friction level: 100
```

**Comment by berarma** [comment]
I guess you're using module `hid-tmff2`. It doesn't support that setting. I have to look why the slider isn't grayed out.

**Comment by kevenwyld** [comment]
>  guess you're using module hid-tmff2

Ah, yes i am.

> It doesn't support that setting

Well, it does though.  At least when I adjust that setting when playing a game it adjusts the overall strength of FFB.  Also when debug is on and I move the slider it does seem to scale 0-100% to 0-65535.  In this log `49151` is equivalent to 75% :

```
DEBUG:root:Setting FF gain: 48495
DEBUG:root:Setting FF gain: 47185
DEBUG:root:Setting FF gain: 43253
DEBUG:root:Setting FF gain: 40631
DEBUG:root:Setting FF gain: 36699
DEBUG:root:Setting FF gain: 34078
DEBUG:root:Setting FF gain: 32112
DEBUG:root:Setting FF gain: 31456
DEBUG:root:Setting FF gain: 30146
DEBUG:root:Setting FF gain: 28835
DEBUG:root:Setting FF gain: 26214
DEBUG:root:Setting FF gain: 22937
DEBUG:root:Setting FF gain: 21626
DEBUG:root:Setting FF gain: 19660
DEBUG:root:Setting FF gain: 18349
DEBUG:root:Setting FF gain: 16383
DEBUG:root:Setting FF gain: 14417
DEBUG:root:Setting FF gain: 11140
DEBUG:root:Setting FF gain: 8519
DEBUG:root:Setting FF gain: 655
DEBUG:root:Setting FF gain: 0
DEBUG:root:Setting FF gain: 45219
DEBUG:root:Setting FF gain: 45874
DEBUG:root:Setting FF gain: 46529
DEBUG:root:Setting FF gain: 47185
DEBUG:root:Setting FF gain: 47840
DEBUG:root:Setting FF gain: 48495
DEBUG:root:Setting FF gain: 49151
```

**Comment by berarma** [comment]
Yes, you're right. There can be two ways of adjusting the FFB gain, I'm using the one available in your driver. I didn't remember that.

The issue here is that I can set the gain value but I can't read it in the same way from the system, that's why the slider it's always at 100 when you start Oversteer.

It's a difficult case and I'm not sure I should be changing the gain using that method. You should be aware that games can override this gain and there's no way for Oversteer to know it.

I'll try to fix it anyway.

**Comment by kevenwyld** [comment]
> You should be aware that games can override this gain and there's no way for Oversteer to know it.

I was not aware of this, thanks for the information.

I suppose knowing the value would be more valuable than trying to change it if that's the case which is unfortunate.  I poked around in the `/sys/devices` tree for the t300rs but I can't seem to find anything that's changing when gain is adjusted.  Unlike `Spring effect gain` for example which corresponds to the `spring_level` device file accepting a value of 0-100.  I'm guessing that's why it's complicated to read that value in oversteer?

Anyway thanks a lot for taking a look at this.  I really appreciate your time and it's helpful to better understand what's happening.

**Comment by berarma** [comment]
> I suppose knowing the value would be more valuable than trying to change it if that's the case which is unfortunate. I poked around in the `/sys/devices` tree for the t300rs but I can't seem to find anything that's changing when gain is adjusted. Unlike `Spring effect gain` for example which corresponds to the `spring_level` device file accepting a value of 0-100. I'm guessing that's why it's complicated to read that value in oversteer?

Yes, that's the missing part. You've understood it very well.

I'll take a look anyway since the current behavior is misleading.

**Comment by Kimplul** [comment]
Sorry to necrobump, but I could probably fairly easily add in a `/sys` entry for global gain.

Just to be sure, this global gain would act as a scalar for whatever gain a game would request, right? Should autocentering also work this way?

**Comment by berarma** [comment]
This setting wasn't originally in the Logitech module I based new-lg4ff on. I added it so the user could have some control in the final gain. The driver uses the gain set through the API multiplied by the gain in `/sys`.

The autocenter value is overriden instead. Looking at it now, I don't understand the reason to implement it in `/sys`. It should be controlled from the API. Games normally don't change this.

I wanted to implement many settings in new-lg4ff when I was testing it so it would help me diagnose issues. Now I'd like to get feedback about the real purpose of these settings. I think the master gain is implemented in higher end wheels at the firmware level. Although less needed in low powered wheels I think it can still be useful.

I'm not sure about autocenter, at least in the way I implemented it in new-lg4ff. It could be useful to make the wheel autocenter without affecting the value set by the API.

In summary, go ahead but maybe don't just implement it because is easy but because it's useful too.

**Comment by Kimplul** [comment]
Alright, I have a preliminary patch of Oversteer over at https://github.com/Kimplul/oversteer and a driver patch in https://github.com/Kimplul/hid-tmff2/tree/t248. I just added a master gain and ignored autocentering for now. On Windows, Thrustmaster has a tickbox for using global autocentering with a fixed value or letting games set the value dynamically, but not both at the same time. I think this is a reasonable approach, but could be worth discussing.

At the moment I'm working on adding in support for the Thrustmaster T248, the driver seems to work but some games (mainly DiRT Rally 2 and DiRT 4) refuse to use FF, not sure why. Anycase, I'll open up a pull request to Oversteer with my patches once I merge `t248` into `main`, but if @kevenwyld would like to test them out before then feel free to.

**Comment by kevenwyld** [comment]
Thanks to both of you for working on this.  I will test as soon as I am able.  I have been struck by thrustmaster's unfortunate reputation for mediocre build quality and my wheel is apart on the bench at the moment because the right paddle button randomly get stuck on.  I will test as soon as I get it going again.

**Comment by berarma** [comment]
> Alright, I have a preliminary patch of Oversteer over at https://github.com/Kimplul/oversteer and a driver patch in https://github.com/Kimplul/hid-tmff2/tree/t248. I just added a master gain and ignored autocentering for now. On Windows, Thrustmaster has a tickbox for using global autocentering with a fixed value or letting games set the value dynamically, but not both at the same time. I think this is a reasonable approach, but could be worth discussing.

That's how I think it should be in retrospect. Does it handle gain in the same way?

> At the moment I'm working on adding in support for the Thrustmaster T248, the driver seems to work but some games (mainly DiRT Rally 2 and DiRT 4) refuse to use FF, not sure why. Anycase, I'll open up a pull request to Oversteer with my patches once I merge `t248` into `main`, but if @kevenwyld would like to test them out before then feel free to.

Codemasters seems to use non-standard wheel detection methods mixed with predefined profiles. I'd try with Proton versions older than 7.0 since they've switched hid implementations and there's some regressions.

**Comment by Kimplul** [comment]
> Does it handle gain in the same way?

No, gain is from what I can tell just a scalar to whatever the game itself sets. The T300RS does set the gain for the wheel with a separate configuration request, but with respect to global gain set in the Thrustmaster configuration utility.

> Codemasters seems to use non-standard wheel detection methods mixed with predefined profiles. You should try with Proton versions older than 7.0 if you haven't already since they've switched hid implementations and there are some regressions.

Apparently the T248 uses the `Slider` HID axis for the clutch, and only Proton 7.0 (and presumably newer) handle this axis 'correctly', which is sort of unfortunate. Still, no FFB with Proton 6 at least, but now with no working clutch.

As an aside, input handling is a lot more complex than I would've thought.
+ Linux parses a HID `Generic Desktop Page` absolute `Slider` axis as `ABS_THROTTLE`, which is passed on to `evdev`
+ Wine internally parses `ABS_THROTTLE` as  `Simulation Page Throttle` from `evdev` and passes it on to `dinput`
+ `dinput` parses `Simulation Page Throttle`, `Generic Desktop Page Slider`, and `Generic Desktop Page Dial` as `GUID_Slider`, which is the correct axes that this particular wheel uses, so games that already have a preset for the T248 work as intended.

Amazing.
